src/Collection/CollectionTrait.php

```diff
@@ -686,6 +686,47 @@ public function _unwrap()
      *
      * @return \Cake\Collection\CollectionInterface
      */
+    public function cartesianProduct(callable $operation = null, callable $filter = null)
+    {
+        if ($this->isEmpty()) {
+            return new Collection([]);
+        }
+
+        $result = [];
+        $counts = $this->map(function ($arr) {
+            return count($arr);
+        })->toList();
+        $allArr = $this->toList();
+        $lastIndex = count($allArr) - 1;
+        // holds the indexes of the arrays that generate the current combination
+        $currentIndexes = array_fill(0, $lastIndex + 1, 0);
+        $changeIndex = $lastIndex;
+
+        while (!($changeIndex === 0 && $currentIndexes[0] === $counts[0])) {
+            $currentCombination = array_map(function ($arr, $index) {
+                return $arr[$index];
+            }, $allArr, $currentIndexes);
+
+            if ($filter === null || $filter(...$currentCombination)) {
+                $result[] = ($operation === null) ? $currentCombination : $operation(...$currentCombination);
```

Comment:
You can't use spread operators in 3.3.x, this will have to target 3.4.x if you want to use the spread operator.